extensions/amp-bind/0.1/bind-expr-impl.jison

```diff
@@ -0,0 +1,288 @@
+"null"                    return 'NULL'
+"NULL"                    return 'NULL'
+"TRUE"                    return 'TRUE'
+"true"                    return 'TRUE'
+"FALSE"                   return 'FALSE'
+"false"                   return 'FALSE'
+[0-9]+("."[0-9]+)?\b      return 'NUMBER'
+[a-zA-Z_][a-zA-Z0-9_]*    return 'NAME'
+\'[^\']*\'                return 'STRING'
+\"[^\"]*\"                return 'STRING'
+.                         return 'INVALID'
+<<EOF>>                   return 'EOF'
+
+/lex
+
+/* token type names (no precedence) */
+
+%token NAME
+%token STRING
+%token NUMBER
+%token TRUE FALSE
+%token NULL
+%token EOF
+
+/*
+ * operator precedence
+ * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence
+ */
+
+%right '?' ':'
+%left '||'
+%left '&&'
+%left '==' '!='
+%left '<' '<=' '>' '>='
+%left '+' '-'
+%left '*' '/' '%'
+%right '!' UMINUS UPLUS
+%left '(' ')'
+%left '.' '[' ']'
+
+%%
+
+/* language grammar */
+
+result:
+    expr EOF
+      {return $1;}
+  | EOF
+      {return '';}
+  ;
+
+expr:
+    operation
+      {$$ = $1;}
+  | invocation
+      {$$ = $1;}
+  | member_access
+      {$$ = $1;}
+  | '(' expr ')'
+      {$$ = $2;}
+  | variable
+      {$$ = $1;}
+  | literal
+      {$$ = $1;}
+  ;
+
+operation:
+    '!' expr
+      {$$ = !$2;}
+  | '-' expr %prec UMINUS
+      {$$ = -$2;}
+  | '+' expr %prec UPLUS
+      {$$ = +$2;}
+  |  expr '+' expr
+      {$$ = $1 + $3;}
+  | expr '-' expr
+      {$$ = $1 - $3;}
+  | expr '*' expr
+      {$$ = $1 * $3;}
+  | expr '/' expr
+      {$$ = $1 / $3;}
+  | expr '%' expr
+      {$$ = $1 % $3;}
+  | expr '&&' expr
+      {$$ = $1 && $3;}
+  | expr '||' expr
+      {$$ = $1 || $3;}
+  | expr '<=' expr
+      {$$ = $1 <= $3;}
+  | expr '<' expr
+      {$$ = $1 < $3;}
+  | expr '>=' expr
+      {$$ = $1 >= $3;}
+  | expr '>' expr
+      {$$ = $1 > $3;}
+  | expr '!=' expr
+      {$$ = $1 != $3;}
+  | expr '==' expr
+      {$$ = $1 == $3;}
+  | expr '?' expr ':' expr
+      {$$ = $1 ? $3 : $5;}
+  ;
+
+invocation:
+    expr '.' NAME args
+      %{
+        $$ = null;
+
+        var obj = Object.prototype.toString.call($1);
+
+        var whitelist = functionWhitelist[obj];
+        if (whitelist) {
+          var fn = $1[$3];
+          if (whitelist.indexOf(fn) >= 0) {
+            if (typeCheckArgs($4)) {
+              $$ = fn.apply($1, $4);
+            } else {
+              throw new Error(`Unexpected argument type in {$3}()`);
+            }
+            return;
+          }
+        }
+
+        throw new Error(`{$3}() is not a supported function.`);
+      %}
+  ;
+
+args:
+    '(' ')'
+      {$$ = [];}
+  | '(' array ')'
+      {$$ = $2;}
+  ;
+
+member_access:
+    expr member
+      %{
+        $$ = null;
+
+        if ($1 === null || $2 === null) {
+          return;
+        }
+
+        var prop = Object.prototype.toString.call($2);
+        if (prop === '[object String]' || prop === '[object Number]') {
```

Comment:
I actually had to revert this change since `hasOwnProperty` throw a "cannot convert to primitive" error in some cases. Changed to use `typeof` instead of `toString` though, which should be faster.